### DISCUSSION TITLE
So much trouble with Supabase - switch to Firebase? (with UPDATE)

### DISCUSSION THREAD

**Discussion by ykdojo** [ykdojo/defaang#139] [General]
Hey everyone!

I've done everything I could do today with Supabase to make it work with the frontend-backend connections, but I'm having a LOT of trouble with it.

More details: https://github.com/ykdojo/defaang/issues/108

Anyway, I'm now seriously considering switching to Firebase.

It's supposedly more mature, better documented, and it's more battle-tested if I understand it correctly.

Any thoughts?

I'm thinking of playing around with it sometime soon to see if I can get it to work.

With it, [NextAuth](https://next-auth.js.org/adapters/firebase) might be a good option, too.

UPDATE:

Based on my convo with @subhoghoshX below, I was able to resolve this issue! I'm thinking perhaps we should give it a try a bit more... Let's see how it goes.

**Comment by RakeshSangem** [comment]
I have worked with Firebase firestore. It's blazingly fast and documentation is top notch in my opinion.

**Comment by subhoghoshX** [comment]
I didn't know that you were facing issues with Supabase. I thought it was straightforward to work with once you figure out the policies. Could you please elaborate on what issues you're facing?

**Reply by ykdojo** [reply]
Sure!

So [here](https://github.com/supabase/supabase/issues/147#issuecomment-687579946), they say the user method takes a JWT - but it doesn't.

And [here](https://supabase.com/docs/reference/javascript/auth-setauth), they explain how to set a JWT in backend, but it doesn't seem to work, either.

Based on those two things and a few other things (mostly the fact that documentation is often incomplete and feels scattered), I feel like it'll be better to switch to Firebase.

**Reply by subhoghoshX** [reply]
The first link is probably outdated as it's from 2020. And I don't quite understand the reason behind using JWT and the NextJS API altogether. Why not go with the straightforward route that they mention in their docs? That is making the requests directly from the client.

**Reply by ykdojo** [reply]
I don't think that would quite work.

For example, say we want to insert a new question to our questions table.

Then, we want to set is_approved = false initially.

But if we let a user insert whatever they want as long as created_by is equal to the current user, they'll be able to set is_approved to true, too.

I guess we could have another table to keep track of the approval mechanism, but I think that'd be too much complexity for this.

**Reply by ykdojo** [reply]
@subhoghoshX actually [this one](https://github.com/supabase/supabase/issues/147#issuecomment-949953227) worked! I looked into it because you mentioned the previous comment was outdated - so I'm willing to try Supabase a bit more haha

**Reply by subhoghoshX** [reply]
haha cool. So we're going the NextJS API route. I wish we could do it the easy way.

There's also something called column-level security and views. I'll try to look into it.

**Reply by ykdojo** [reply]
I think using the NextJS route is pretty simple actually - simple enough, anyway.

I'll send a PR when it's ready

**Reply by subhoghoshX** [reply]
That is true. It's quite easy. My only concern was going through API would increase latency.

**Reply by ykdojo** [reply]
The latency we should worry about is with views that have many things in it - so maybe a home view with many questions listed, or a question view with many answers attached to it.

As long as we do server-side render most of them, I figured it'd be ok.

I think it's ok to have slight latency on less-used functionalities like the submission box.

**Reply by subhoghoshX** [reply]
Yeah definitely. Or we can just use loaders. Using loaders is a pretty common pattern.

**Reply by ykdojo** [reply]
Interesting. Any reference article, post or site we can look at for that?

**Reply by subhoghoshX** [reply]
Thanks for asking. I never thought of doing research into it. I just always liked them by instinct.

Here by loaders, I also mean skeleton loaders as well. I personally prefer skeletons much more than spinners. There are many websites using them like Supabase, LeetCode, Quora, YouTube, etc.

I found this [article](https://uxdesign.cc/what-you-should-know-about-skeleton-screens-a820c45a571a) and this [talk](https://www.youtube.com/watch?v=EbbjEY-TyhU&t=4136s) to be quite interesting. And this Reddit [post](https://www.reddit.com/r/iOSProgramming/comments/nniqoe/its_amazing_how_much_of_a_difference_a_skeleton/) as well.

**Reply by ykdojo** [reply]
Thanks! The skeleton loader in the Reddit post does look really cool

**Reply by subhoghoshX** [reply]
Here is an example [skeleton loader](https://play.tailwindcss.com/u2rPeQkPOW) implemented with pure tailwindcss. We can do something like this in some parts of the UI.
